pkg/kv/kvserver/raftlog/iterator.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package raftlog
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/util/iterutil"
+	"github.com/cockroachdb/errors"
+)
+
+type storageIter interface {
+	SeekGE(key storage.MVCCKey)
+	Valid() (bool, error)
+	Next()
+	Close()
+	UnsafeValue() []byte
+}
+
+// Reader is the subset of storage.Reader relevant for accessing the raft log.
+//
+// The raft log is a contiguous sequence of indexes (i.e. no holes) which may be
+// empty.
+type Reader interface {
+	NewMVCCIterator(storage.MVCCIterKind, storage.IterOptions) storage.MVCCIterator
+}
+
+// An Iterator inspects the raft log. After creation, SeekGE should be invoked,
+// followed by Next (and SeekGE again if desired, etc). The iterator must no
+// longer be used after any of its method has returned an error. The Iterator
+// must eventually be Close'd to release the resources associated to it.
+type Iterator struct {
+	eng       Reader
+	prefixBuf keys.RangeIDPrefixBuf
+
+	iter storageIter
+	// TODO(tbg): we're not reusing memory here. Since all of our allocs come
+	// from protobuf marshaling, this is hard to avoid but we can do a little
+	// better and at least avoid a few allocations.
+	entry *Entry
+}
+
+// IterOptions are options to NewIterator.
+type IterOptions struct {
+	// Hi ensures the Iterator never seeks to any Entry with index >= Hi. This is
+	// useful when the caller is interested in a slice [Lo, Hi) of the raft log.
+	Hi uint64
```

Comment:
This probably is a result of `MVCCIterator` interface, but it's a bit asymmetrical that the iterator has to be seeked to the `Lo` position, whereas the `Hi` position can be specified at the creation time.

Would it make sense to have an API where you set both at creation?

Can you and do you ever need to re-`Seek` an iterator to a lower position? Higher position? Is there a benefit (performance?) in reusing iterators?

If there is low benefit in reusing, I could imagine a simpler / more restictive API:
- `NewMVCCIterator` / `NewIterator` return an iterator with defined `[Lo,Hi)` straight away.
- It's a "forward" iterator, so only has `Next`, doesn't have `Seek`, or allows only seeking forward.

Reply:
Given that the only user of this is `Visit`, the above should make sense (side note: could even unexport it for now?). Are there other potential users of this iterator? Is the common denominator still the scan-once pattern that `Visit` does?

Reply:
There was an earlier iteration of the Iterator as well, the current form is after suggestions from the storage team. I think we can iterate on this more in the future but I'd like to get this PR merged if there are no real objections.

Reply:
As per a DM chat, left TODO comment on Iterator.